// @aws-cdk/aws-rds/test/integ.cluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ec2 = require("@aws-cdk/aws-ec2");
const kms = require("@aws-cdk/aws-kms");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../lib");
const parameter_group_1 = require("../lib/parameter-group");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-rds-integ');
const vpc = new ec2.Vpc(stack, 'VPC', { maxAzs: 2 });
const params = new parameter_group_1.ClusterParameterGroup(stack, 'Params', {
    family: 'aurora5.6',
    description: 'A nice parameter group',
    parameters: {
        character_set_database: 'utf8mb4',
    },
});
const kmsKey = new kms.Key(stack, 'DbSecurity');
const cluster = new lib_1.DatabaseCluster(stack, 'Database', {
    engine: lib_1.DatabaseClusterEngine.AURORA,
    masterUser: {
        username: 'admin',
        password: cdk.SecretValue.plainText('7959866cacc02c2d243ecfe177464fe6'),
    },
    instanceProps: {
        instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE3, ec2.InstanceSize.SMALL),
        vpcSubnets: { subnetType: ec2.SubnetType.PUBLIC },
        vpc,
    },
    parameterGroup: params,
    kmsKey,
});
cluster.connections.allowDefaultPortFromAnyIpv4('Open to the world');
app.synth();
