// @aws-cdk/aws-rds/test/test.parameter-group.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'create a parameter group'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new lib_1.ParameterGroup(stack, 'Params', {
            family: 'hello',
            description: 'desc',
            parameters: {
                key: 'value',
            },
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::RDS::DBParameterGroup', {
            Description: 'desc',
            Family: 'hello',
            Parameters: {
                key: 'value',
            },
        }));
        test.done();
    },
    'create a cluster parameter group'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new lib_1.ClusterParameterGroup(stack, 'Params', {
            family: 'hello',
            description: 'desc',
            parameters: {
                key: 'value',
            },
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::RDS::DBClusterParameterGroup', {
            Description: 'desc',
            Family: 'hello',
            Parameters: {
                key: 'value',
            },
        }));
        test.done();
    },
    'Add an additional parameter to an existing parameter group'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const clusterParameterGroup = new lib_1.ClusterParameterGroup(stack, 'Params', {
            family: 'hello',
            description: 'desc',
            parameters: {
                key1: 'value1',
            },
        });
        clusterParameterGroup.addParameter('key2', 'value2');
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::RDS::DBClusterParameterGroup', {
            Description: 'desc',
            Family: 'hello',
            Parameters: {
                key1: 'value1',
                key2: 'value2',
            },
        }));
        test.done();
    },
};
