// @aws-cdk/aws-rds/test/test.database-secret.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../lib");
module.exports = {
    'create a database secret'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        // WHEN
        new lib_1.DatabaseSecret(stack, 'Secret', {
            username: 'admin-username',
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::SecretsManager::Secret', {
            Description: {
                'Fn::Join': [
                    '',
                    [
                        'Generated by the CDK for stack: ',
                        {
                            Ref: 'AWS::StackName',
                        },
                    ],
                ],
            },
            GenerateSecretString: {
                ExcludeCharacters: '"@/\\',
                GenerateStringKey: 'password',
                PasswordLength: 30,
                SecretStringTemplate: '{"username":"admin-username"}',
            },
        }));
        test.done();
    },
    'with master secret'(test) {
        // GIVEN
        const stack = new core_1.Stack();
        const masterSecret = new lib_1.DatabaseSecret(stack, 'MasterSecret', {
            username: 'master-username',
        });
        // WHEN
        new lib_1.DatabaseSecret(stack, 'UserSecret', {
            username: 'user-username',
            masterSecret,
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::SecretsManager::Secret', {
            GenerateSecretString: {
                ExcludeCharacters: '"@/\\',
                GenerateStringKey: 'password',
                PasswordLength: 30,
                SecretStringTemplate: {
                    'Fn::Join': [
                        '',
                        [
                            '{"username":"user-username","masterarn":"',
                            {
                                Ref: 'MasterSecretA11BF785',
                            },
                            '"}',
                        ],
                    ],
                },
            },
        }));
        test.done();
    },
};
